src/common/dom/apply_themes_on_element.ts

```diff
@@ -33,17 +19,53 @@ let PROCESSED_THEMES: { [key: string]: ProcessedTheme } = {};
 export const applyThemesOnElement = (
   element,
   themes: HomeAssistant["themes"],
-  selectedTheme?: string
+  selectedTheme?: string,
+  themeOptions?: Partial<HomeAssistant["selectedTheme"]>
 ) => {
-  const newTheme = selectedTheme
-    ? PROCESSED_THEMES[selectedTheme] || processTheme(selectedTheme, themes)
-    : undefined;
+  let cacheKey = selectedTheme;
+  let themeRules: Partial<Theme> | undefined;
 
-  if (!element._themes && !newTheme) {
+  if (selectedTheme === "default" && themeOptions) {
+    if (themeOptions.dark) {
+      cacheKey = `${cacheKey}__dark`;
+      themeRules = { ...themeRules, ...darkStyles };
+    }
+    if (themeOptions.primaryColor) {
+      cacheKey = `${cacheKey}__primary_${themeOptions.primaryColor}`;
+      const primaryColor = Color(themeOptions.primaryColor);
+      themeRules = {
+        ...themeRules,
+        "primary-color": themeOptions.primaryColor,
+        "light-primary-color": primaryColor.lighten(0.5).hex(),
+        "dark-primary-color": primaryColor.darken(0.5).hex(),
+        "text-primary-color":
+          primaryColor.contrast(Color("#ffffff")) > 2.6 ? "#fff" : "#212121",
+        "state-icon-color": primaryColor.desaturate(0.5).hex(),
+      };
+    }
+    if (themeOptions.accentColor) {
+      cacheKey = `${cacheKey}__accent_${themeOptions.accentColor}`;
+      themeRules = {
+        ...themeRules,
```

Comment:
Do we really need to copy it all the time? Can't we just start with a copy and set it direectly ?